src/OpenLoco/src/GameSaveCompare.cpp

```diff
@@ -0,0 +1,385 @@
+#include "GameSaveCompare.h"
+
+#include <string>
+
+#include "Effects/EffectsManager.h"
+#include "GameState.h"
+#include "Logging.h"
+#include "OpenLoco.h"
+#include "S5/Limits.h"
+#include "S5/S5.h"
+#include "Vehicles/Vehicle.h"
+#include <OpenLoco/Interop/Interop.hpp>
+
+using namespace OpenLoco::Diagnostics;
+
+namespace OpenLoco::GameSaveCompare
+{
+    std::string getVehicleSubType(const Vehicles::VehicleEntityType vehicleSubType);
+    std::string getEffectSubType(const EffectType effectSubType);
+    void logVehicleTypeAndSubTYpe(int offset, const OpenLoco::Entity& entity);
+    void logEffectType(int offset, const OpenLoco::Entity& entity);
+    void logDivergentEntityOffset(const OpenLoco::Entity& lhs, const S5::Entity& rhs, int offset);
+    void compareGameStates(GameState& gameState1, S5::GameState& gameState2);
+    void compareElements(const std::vector<S5::TileElement>& tileElements1, const std::vector<S5::TileElement>& tileElements2);
+
+    namespace unsafe
+    {
+        template<typename T>
+        constexpr auto begin(const T& item)
+        {
+            return reinterpret_cast<const char*>(&item);
+        }
+        template<typename T>
+        constexpr auto end(const T& item)
+        {
+            return reinterpret_cast<const char*>(&item) + sizeof(T);
+        }
+        template<typename T>
+        auto bitWiseEqual(const T& lhs, const T& rhs)
+        {
+            return std::equal(begin(lhs), end(lhs), // will become constexpr with C++20
+                              begin(rhs),
+                              end(rhs));
+        }
+        template<typename T1, typename T2>
+        auto bitWiseEqual(const T1& lhs, const T2& rhs)
+        {
+            return std::equal(begin(lhs), end(lhs), // will become constexpr with C++20
+                              begin(rhs),
+                              end(rhs));
+        }
+
+        template<typename T1, typename T2>
+        void bitWiseLogDivergence(const std::string type, const T1& lhs, const T2& rhs, bool printAHeader)
+        {
+            size_t size = sizeof(T1) / sizeof(char);
+            char* array_lhs = (char*)(&lhs);
+            char* array_rhs = (char*)(&rhs);
```

Comment:
You could use `getSpanBytes` here once that function exists which also provides `size()`

Reply:
It should also probably have a static assert to ensure T1 and T2 are always of the same size.